src/main/java/org/apache/commons/lang3/CharSequenceUtils.java

```diff
@@ -133,24 +133,87 @@ static int indexOf(final CharSequence cs, final int searchChar, int start) {
-//            //       If so then calling toString() on it is better than reverting to
-//            //       the green implementation in the else block
-//            return ((String) cs).indexOf((String) searchChar, start);
-//        } else {
-//            // TODO: Implement rather than convert to String
-//            return cs.toString().indexOf(searchChar.toString(), start);
-//        }
+    static int indexOf(final CharSequence cs, final CharSequence searchChar, int start) {
+        // if searchChar is a String, and cs is some specific kind of AbstractString,
+        //   which both have a indexOf function and widely used by normal java codes,
+        // then we just invoke their indexOf function.
+        if (searchChar instanceof String) {
+            if (cs instanceof String) {
+                return ((String) cs).indexOf((String) searchChar, start);
+            } else if (cs instanceof StringBuilder) {
+                return ((StringBuilder) cs).indexOf((String) searchChar, start);
+            } else if (cs instanceof StringBuffer) {
+                return ((StringBuffer) cs).indexOf((String) searchChar, start);
+            }
+        }
+
+        int len1 = cs.length();
+        int len2 = searchChar.length();
+
+        // successful-result >= 0
+        if (start < 0) {
+            start = 0;
+        }
+
+
+        // successful-result <= len1
+        if (start > len1) {
+            start = len1;
+        }
+
+        // if len2 == 0 return directly.
+        if (len2 == 0) {
+            return start;
+        }
+
+        // limit means the largest possible value of successful-result
+        final int limit = len1 - len2;
+
+        // if start > limit, then have no enough length for a search,
+        //   thus cannot find a index, thus fail.
+        // if len2 < 0, then it is illegal for this function.
+        // if limit < 0, then it means len2 > len1, thus fail.
+        if (start > limit || len2 < 0 || limit < 0) {
+            return -1;
+        }
+
+        // notice that when we enter here, we make sure:
+        // start in [0, limit]
+        // limit in [0, len1-1]
+        // len2 in [1, len1]
+
+        // if len2 is small enough, and cs is some specific kind of AbstractString,
+        //   which both have a indexOf function and widely used by normal java codes,
+        // then we just invoke their indexOf function.
+        if (len2 <= TO_STRING_LIMIT) {
+            if (cs instanceof String) {
+                return ((String) cs).indexOf(searchChar.toString(), start);
+            } else if (cs instanceof StringBuilder) {
+                return ((StringBuilder) cs).indexOf(searchChar.toString(), start);
+            } else if (cs instanceof StringBuffer) {
+                return ((StringBuffer) cs).indexOf(searchChar.toString(), start);
+            }
+        }
+
+        char char0 = searchChar.charAt(0);
+
+        int i = start;
+        while (true) {
```

Comment:
@aherbert 
I do think it will run faster the way I wrote it...
However you can prove me wrong, means your version be faster, or equal speed in jmh, with most cases, then I'm glad to refine it.

Reply:
Note that the `&&` condition is only evaluated on the right if the left is true thus you effectively have:
```java
for (int i = start; i <= limit; i++) {
    if (cs.charAt(i) == char0) {
        // ... more checks
    }
}
```

My suggestion was for code clarity:
```
for i in [start, end]
     if match at position i
          return i

return -1
```
It is simpler. If you think your way is faster then that remains to be proven.

Reply:
> If you think your way is faster then that remains to be proven.

OK, I will write&run a jmh now.
If it shows you be correct, then I'm glad to change the codes.

Reply:
@aherbert
yep, you are correct.
glad to fix it.